File: 94. CourseScheduleII.cpp
#include <iostream>
#include <vector>
#include <queue>
using namespace std;

vector<int> findOrder(int numCourses, vector<vector<int>>& prerequisites) {
    vector<vector<int>> graph(numCourses);
    vector<int> indegree(numCourses, 0);

    for (auto& prerequisite : prerequisites) {
        int course = prerequisite[0];
        int required = prerequisite[1];
        graph[required].push_back(course);
        indegree[course]++;
    }

    queue<int> q;
    for (int course = 0; course < numCourses; course++) {
        if (indegree[course] == 0) q.push(course);
    }

    vector<int> order;
    while (!q.empty()) {
        int current = q.front();
        q.pop();
        order.push_back(current);
        for (int next : graph[current]) {
            indegree[next]--;
            if (indegree[next] == 0) q.push(next);
        }
    }

    if ((int)order.size() != numCourses) return {};
    return order;
}

void printVector(const vector<int>& v) {
    cout << "[";
    for (size_t i = 0; i < v.size(); i++) { if (i) cout << ", "; cout << v[i]; }
    cout << "]\n";
}

int main() {
    vector<vector<int>> p1 = {{1, 0}};
    printVector(findOrder(2, p1));
    vector<vector<int>> p2 = {{1, 0}, {2, 0}, {3, 1}, {3, 2}};
    printVector(findOrder(4, p2));
    vector<vector<int>> p3 = {};
    printVector(findOrder(1, p3));
    vector<vector<int>> p4 = {{1, 0}, {0, 1}};
    printVector(findOrder(2, p4));
    return 0;
}

File: 94. CourseScheduleII.java
import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Queue;

public class CourseScheduleII {

    public static int[] findOrder(int numCourses, int[][] prerequisites) {
        List<List<Integer>> graph = new ArrayList<>();
        int[] indegree = new int[numCourses];

        for (int i = 0; i < numCourses; i++) {
            graph.add(new ArrayList<>());
        }

        for (int[] prerequisite : prerequisites) {
            int course = prerequisite[0];
            int required = prerequisite[1];
            graph.get(required).add(course);
            indegree[course]++;
        }

        Queue<Integer> queue = new ArrayDeque<>();
        for (int course = 0; course < numCourses; course++) {
            if (indegree[course] == 0) {
                queue.offer(course);
            }
        }

        int[] order = new int[numCourses];
        int index = 0;

        while (!queue.isEmpty()) {
            int current = queue.poll();
            order[index++] = current;

            for (int next : graph.get(current)) {
                indegree[next]--;
                if (indegree[next] == 0) {
                    queue.offer(next);
                }
            }
        }

        if (index != numCourses) {
            return new int[0];
        }

        return order;
    }

    public static void main(String[] args) {
        System.out.println(Arrays.toString(findOrder(2, new int[][]{{1, 0}})));
        System.out.println(Arrays.toString(findOrder(4, new int[][]{{1, 0}, {2, 0}, {3, 1}, {3, 2}})));
        System.out.println(Arrays.toString(findOrder(1, new int[][]{})));
        System.out.println(Arrays.toString(findOrder(2, new int[][]{{1, 0}, {0, 1}})));
    }
}
